frontend/annotation_resolver.py

```diff
@@ -60,82 +35,138 @@ def resolve(self, annotation, solver):
-            assert len(args_and_return) == 2
-            assert args_and_return[0][0] == "[" and args_and_return[0][-1] == "]"
-
-            # Get the args and return types
-            args_annotations = self.resolve_args(args_and_return[0][1:-1])
-            args_types = [self.resolve(arg, solver) for arg in args_annotations]
-            return_type = self.resolve(args_and_return[1], solver)
-
-            return self.z3_types.funcs[len(args_types)](*(args_types + [return_type]))
-
-        if annotation[:5] == "Union":
-            # Parse Union type
-            assert annotation[5] == "[" and annotation[-1] == "]"
-
-            # Get the types of the union args
-            args_annotations = self.resolve_args(annotation[6:-1])
-            args_types = [self.resolve(arg, solver) for arg in args_annotations]
-
-            # The result of the union type is only one of args, Z3 picks the appropriate one
-            # according to the added constraints.
-            # Therefore, using more than one type from the union in the same program isn't yet supported.
-            # For example, the following is not supported:
-            # def f(x: Union[int, str]): ...
-            # f(1)
-            # f("str")
-            result_type = solver.new_z3_const("union")
-            solver.add(Or([result_type == arg for arg in args_types]),
-                       fail_message="Union in type annotation")
-
+        if isinstance(annotation, ast.Name):
+            if annotation.id in self.primitives:
+                return self.primitives[annotation.id]
+            if annotation.id in self.z3_types.all_types:
+                return getattr(self.z3_types.type_sort, "class_{}".format(annotation.id))
+            
+            # Check if it's a generic type var
+            if generics_map is None:
+                raise ValueError("Invalid type annotation {} in line {}".format(annotation.id, annotation.lineno))
+            if annotation.id in generics_map:
+                return generics_map[annotation.id]
+            if annotation.id not in self.type_vars:
+                raise ValueError("Invalid type annotation {} in line {}".format(annotation.id, annotation.lineno))
+
+            result_type = solver.new_z3_const("generic")
+            generics_map[annotation.id] = result_type
+
+            supers_types = [self.resolve(x, solver, generics_map) for x in self.type_vars[annotation.id]]
+            if supers_types:
+                solver.add(Or([solver.z3_types.subtype(result_type, x) for x in supers_types]),
+                           fail_message="Generic type in line {}".format(annotation.lineno))
             return result_type
 
-        raise ValueError("Invalid type annotation: {}.".format(annotation))
+        if isinstance(annotation, ast.Subscript):
+            assert isinstance(annotation.value, ast.Name)
+            assert isinstance(annotation.slice, ast.Index)
```

Comment:
This assertion will fail when someone actually attempts slicing. Can we not support that? If not, we should at least give them a proper error message.